### PR TITLE
ENH: Add fundamental attenuation calculations.

### PR DESCRIPTION
**Pull request by nguy**
This is an experimental addition of some fundamental calculations. This first file indicates the rough structure of the files to included (6 or 7 total at the end). Let me know if this looks okay in terms of structure, etc. for CSU toolkit.

It would also take modifying the setup.py to initialize them. One possibility is to create a directory called **fundamentals** or such and putting these all under that heading. I have no strong feelings on either direction.

### CODE REVIEW

**Comment by nguy** [comment]
I should also explicitly note to not accept until the rest is added if you decide to move forward.

**Comment by tjlang** [comment]
Thanks, I'll take a look at this soon.

File: csu_radartools/attenuation.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    Notes
+    -----
+    An example from Battan (1973) is for water at 0C m=7.14-2.89j for a
+       wavelength of 3.21 cm and for ice m=1.78-0.0024j for
+       wavelength range from 1-10 cm.
+    See Battan (1973) Ch.4 , Tables 4.1 and 4.2 for values from
+       Gunn and East (1954).
+    Also see Doviak and Zrnic (1993), Fig. 3.3 caption.
+    """
+    Km = (m**2 - 1) / (m**2 + 2)
```
Comment (tjlang) [inline csu_radartools/attenuation.py]: Make this a separate function since it is repeated in scat_coeff as well.

**Comment by tjlang** [comment]
Looks like a good start. I made on comment on the code itself. I think the <i>__init__.py</i> file in this directory should not need to be changed. However, you will need to update the README and the documentation in the setup.py file so that users know this sub-module exists. I might recommend that this all be kept in a single sub-module even if the content expands beyond attenuation. That may require a name change and/or a separate sub-directory (with associated <i>__init__.py</i>) to house everything.

**Comment by nguy** [comment]
How about one large file that is called ```fundamentals.py```?

In addition, I have a conversions that I'd like to rename common.py and add the ```_check_array``` function into. Are you okay with that?

File: csu_radartools/attenuation.py (hunk 1 above)
Comment (nguy) [inline csu_radartools/attenuation.py]: Can do!

**Comment by tjlang** [comment]
Yeah, that works. I can make use of `\_check_array` elsewhere too.

**Comment by nguy** [comment]
@tjlang lot's of changes, they were already staged. See what you think.

Also, should I roll the rev number?

**Comment by nguy** [comment]
Oh and two more things:
1. I'm open to changing names of any functions etc.
2. I didn't connect the ```linearize``` or ```_check_for_array``` functions in existing code yet

File: README.md [hunk 2]
```diff
@@ -36,10 +36,10 @@ routines have been substantially altered.</b>
 
 To access, use the following in your analysis code:
 ```
-from csu_radartools import (csu_fhc, csu_liquid_ice_mass, csu_blended_rain, csu_dsd, 
-                            csu_kdp, csu_misc)
+from csu_radartools import (csu_fhc, csu_liquid_ice_mass, csu_blended_rain, csu_dsd,
+                            csu_kdp, csu_misc, fundamentals, common)
```
Comment (tjlang) [inline README.md]: Feel that `common` should be an internal import, not really an externally advertised module. The `fundamentals` should be the public face of what you are doing.

File: csu_radartools/common.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    if not len_flag:
+        dz = np.array([dz])
+        kdp = np.array([kdp])
+        zdr = np.array([zdr])
+    return dz, zdr, kdp, len_flag
+
+########################
+##  Unit Conversions  ##
+########################
+def dbz2z(dbz):
```
Comment (tjlang) [inline csu_radartools/common.py]: Two line breaks before a `def` declaration, according to pep8 (a standard `csu_radartools` meets).

File: csu_radartools/common.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    """
+    Convert from log [dBZ] to linear Z [mm^6 m^−3] units.
+
+    Parameters
+    ----------
+    dBZ : float or array
+        logarithmic reflectivity value
+    """
+    return 10.**(np.asarray(dbz)/10.)
+
```
Comment (tjlang) [inline csu_radartools/common.py]: Same, two line breaks.

File: csu_radartools/fundamentals.py [hunk 5]
```diff
@@ -0,0 +1,1041 @@
...
+speed_of_light = 3e8 # Speed of light [m/s]
+kBoltz = 1.381e-23 # Boltzmann's constant [ m^2 kg s^-2 K^-1]
+earth_radius = 6371000 # Earth's average radius [m] assuming sphericity
+r43 = earth_radius * 4./3. # 4/3 Approximation effective radius for standard atmosphere [m]
+# Earth radius according to International Union of Geodesy and Geophysics
+
+##############
+##  System  ##
+##############
+
```
Comment (tjlang) [inline csu_radartools/fundamentals.py]: Two line breaks needed.

File: csu_radartools/fundamentals.py [hunk 6]
```diff
@@ -0,0 +1,1041 @@
...
+    bw_hr = np.deg2rad(bw_h)
+    bw_vr = np.deg2rad(bw_v)
+
+    # Calculate the numerator
+    Numer = (np.pi**3 * speed_of_light * power_t * gainlin**2 * tau *
+             bw_hr * bw_vr * alosslin * rlosslin)
+
+    # Calculate the denominator
+    Denom = 1024. * np.log(2) * lam**2
+    return Numer/Denom
```
Comment (tjlang) [inline csu_radartools/fundamentals.py]: Yeah, seeing some pep8 issues in these files, please clean up.

File: csu_radartools/fundamentals.py [hunk 7]
```diff
@@ -0,0 +1,1041 @@
...
+        noiset = 10. * np.log10(noise/10**-3)
+    else:
+        print("Units must be in 'W' or 'dBm'")
+        noiset = np.nan
+    return noiset
+
+###############
+##  Doppler  ##
+###############
+
```
Comment (tjlang) [inline csu_radartools/fundamentals.py]: 2 lines

File: csu_radartools/fundamentals.py [hunk 8]
```diff
@@ -0,0 +1,1041 @@
...
+        Nyquist Velocity [m/s] or Maximum unambiguous range [m]
+    lam : float
+        Radar wavelength [m]
+    """
+    return (speed_of_light * lam / 8.) / np.asarray(varin)
+
+    ########################
+    ##  Mobile Platforms  ##
+    ########################
+
```
Comment (tjlang) [inline csu_radartools/fundamentals.py]: Two lines

File: setup.py [hunk 9]
```diff
@@ -60,7 +60,7 @@
           To access, use the following in your analysis code:
           from csu_radartools import (
               csu_fhc, csu_kdp, csu_dsd, csu_liquid_ice_mass, csu_misc,
-              csu_blended_rain)
+              csu_blended_rain, fundamentals, common)
```
Comment (tjlang) [inline setup.py]: Again, feel that common should be an internal import only, not externally advertised.

**Comment by tjlang** [comment]
Looking good. This is almost there. We need to decide what to do with the `common` module. I feel it should not be externally advertised, only used internally. Or perhaps the whole thing should be merged with `fundamentals`. Another thought is we only put functions in it that are used directly in other modules, but this may require some reshuffling between `fundamentals` and `common`. I'm OK with merging both now, as long as we don't advertise `common` in the docs for now. We can figure out its ultimate role later. Also, I can work in the internal imports (e.g., `linearize`, `zdp`, etc.) later.

Other than that, I'd recommend some pep8 cleanup. Also, please consider adding a section on `fundamentals` to the demo notebook. There's too much in there to do a thorough review, but maybe you could do a brief demo of 2-3 functions in it. Just put it after the `csu_misc` section.

Note - I've updated the notebook recently so you'll need to merge the latest pull before you modify.

Sound OK?

Thanks a bunch, this looks to be a great addition!

File: README.md (hunk 2 above)
Comment (nguy) [inline README.md]: You got it.

**Comment by nguy** [comment]
Oops, sorry about the pep8. I meant to run that, but forgot. It passes pep8 on my system now. Hopefully this works better.

I agree with common being removed from the docs and outside. I originally had this as a ```conversion.py```, which ultimately could be put in place just like fundamentals I think. But that can be decided later as you noted. 

Examples added to notebook.

**Comment by tjlang** [comment]
Looks good. Merging, and then we can fine tune from there.

**Pull request merged**